File: Week 01/id_377/189.rotate-array.js
// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var rotate = function(nums, k) {
    // ------------------------- 解法1 start --------------------------
    // 最好记  👍 👍 👍 👍 👍
    // 最优解  😘 😘 😘 😘 😘
    nums.unshift(...nums.splice(nums.length - k));
    // nums.unshift(...nums.splice(nums.length-k, k))

    // ------------------------- 解法2 start --------------------------
    // 最好记  👍 👍 👍 👍
    // 最优解  😘 😘 😘 😘
    // for (let i = 0; i < k; i++) {
    //     nums.unshift(nums.pop());
    // }

    // ------------------------- 解法3 start --------------------------
    // 最好记  👍 👍 👍
    // 最优解  😘 😘 😘 😘
    // nums.splice(0,0,...nums.splice(nums.length-k, k))
    
};
// @lc code=end

console.log(
    rotate([1,2,3,4,5,6,7],3)
);

File: Week 01/id_722/189-rotate-array.go
package main
import "fmt"
func main() {
    nums := []int {1,2,3,4,5,6,7}
	rotate(nums, 3)
	fmt.Println("#v", nums)
	
}

func rotate(slice []int, k int) {
	l := len(slice)
	if l == 0 {
		return 
	}
	for i := 0; i < k; i++ {
		move := slice[l-1]
		for j := l-1; j-1 >= 0; j-- {
			// 后面的移动到前面
			// 6 = 5 
			// 5 = 4
			slice[j] = slice[j-1]
		}
		slice[0] = move
	}
}
